### ISSUE TITLE
large number of anon user accounts are being created

### ISSUE THREAD

**Issue by misaugstad** [ProjectSidewalk/SidewalkWebpage#1455]
I think there is some redirect that is happening that is causing a bunch of anon user accounts to be created in quick succession. When I loaded the page for the first time with a fresh database on my local machine, I think I ended up with like 20 anon users. And sometimes when I run the code on my local machine, my browser will complain that there has been too many redirects, and it prevents the page from loading (I then go to an incognito window and it works).

This seems to be a problem on the server too, b/c I see we have 368 users on the newberg-test server right now, and I really don't think we created that many users while testing.

This needs to be fixed before the deployment to the production server for Newberg. A bunch of fake anon accounts will make data analysis later on very hard.

**Comment by jonfroehlich** [comment]
High priority.

**Comment by misaugstad** [comment]
Here is how I have been able to reproduce:
1. Create a fresh database that has no users (possibly using the schema for Docker that @athersharif made)
1. Try to load any page that isn't the /signUp or /signIn page. I am redirected over and over until my browser stops the redirects, so I end up with 20 new anon users.
1. Reload the page and I get another 20 anon users.

How I stop the loop:
1. Go to the /signUp endpoint and create a new user.
1. This will then try to send you to the landing page and the redirect loop will start again...
1. Go to the /signIn endpoint and sign in with the user you just made.

After a user is successfully made and signed in, I have been unable to reproduce the bug. Even if I sign out, close the browser, use an incognito window, use a different browser, whatever. Once a user has been created and signed in ever, the bug doesn't seem to happen again... Even if I create a new user at the /signUp endpoint, the user is successfully created and signed in; the redirect never seems to happen again :confused: 

It isn't a _solution_ really, but for now I am just going to make sure to use this method to make sure we don't have extra anon user accounts in our databases as I create them... I haven't been able to find anything about this when searching the web :cry: 

If your browser ever complains about "too many redirects", please post in this issue thread, because it is probably the same problem!

**Comment by misaugstad** [comment]
Okay I just tested a bit more, and I think this is really related to dev environment problems. It seems like what is happening is that my browser is logged into some account. Then when I change the database on the back-end to a database where I don't have identical credentials, I end up in an infinite loop, stemming from the discrepancy b/w the server and browser. But if I create an account in each of the databases with the same credential, I can switch between them freely.

This is not a problem in a production environment b/c the server is persistent. And since each database has its own domain (sidewalk-test, sidewalk-newberg, etc), the browser doesn't incorrectly think it is logged into some account that doesn't exist.

I had only seen this happen on the remote server once, and that was when we replaced the newberg database with a new one (b/c we modified what the neighborhoods looked like). In that case, my browser still thought I was logged in, but that account didn't exist in the new database that Jason had just loaded.

In conclusion, we just have to be careful when creating new databases, but this isn't a bug we should expect to see in a production environment. As such, I am planning to close this out for now. (luckily it is always available as a reference)

**Comment by misaugstad** [comment]
Okay looks like this is still happening and I'm not sure what the cause is. Noted again in #1559 and can confirm that this is happening. Here is how I am identifying users that were created that shouldn't exist:

```
SELECT * FROM sidewalk_user
WHERE email LIKE 'anonymous@%'
    AND username <> 'anonymous'
    AND sidewalk_user.user_id NOT IN (
        SELECT sidewalk_user.user_id FROM sidewalk_user
        INNER JOIN webpage_activity ON sidewalk_user.user_id = webpage_activity.user_id
        WHERE email LIKE 'anonymous@%'
            AND username <> 'anonymous'
            AND activity LIKE 'Visit_%'
    );
```

And thus we can delete all their login info like so:
```
-- delete user_role
DELETE
FROM user_role
WHERE user_id IN (
    SELECT sidewalk_user.user_id FROM sidewalk_user
    WHERE email LIKE 'anonymous@%'
        AND username <> 'anonymous'
        AND sidewalk_user.user_id NOT IN (
            SELECT sidewalk_user.user_id FROM sidewalk_user
            INNER JOIN webpage_activity ON sidewalk_user.user_id = webpage_activity.user_id
            WHERE email LIKE 'anonymous@%'
                AND username <> 'anonymous'
                AND activity LIKE 'Visit_%'
        )
);

-- delete user_password_info
DELETE
FROM user_password_info
    USING login_info
WHERE user_password_info.login_info_id = login_info.login_info_id
    AND provider_key IN (
        SELECT sidewalk_user.email FROM sidewalk_user
        WHERE email LIKE 'anonymous@%'
            AND username <> 'anonymous'
            AND sidewalk_user.user_id NOT IN (
                SELECT sidewalk_user.user_id FROM sidewalk_user
                INNER JOIN webpage_activity ON sidewalk_user.user_id = webpage_activity.user_id
                WHERE email LIKE 'anonymous@%'
                    AND username <> 'anonymous'
                    AND activity LIKE 'Visit_%'
            )
    );

-- delete user_login_info
DELETE
FROM user_login_info
WHERE user_id IN (
    SELECT sidewalk_user.user_id FROM sidewalk_user
    WHERE email LIKE 'anonymous@%'
        AND username <> 'anonymous'
        AND sidewalk_user.user_id NOT IN (
            SELECT sidewalk_user.user_id FROM sidewalk_user
            INNER JOIN webpage_activity ON sidewalk_user.user_id = webpage_activity.user_id
            WHERE email LIKE 'anonymous@%'
                AND username <> 'anonymous'
                AND activity LIKE 'Visit_%'
        )
);

-- delete login_info
DELETE
FROM login_info
WHERE provider_key IN (
    SELECT sidewalk_user.email FROM sidewalk_user
    WHERE email LIKE 'anonymous@%'
        AND username <> 'anonymous'
        AND sidewalk_user.user_id NOT IN (
            SELECT sidewalk_user.user_id FROM sidewalk_user
            INNER JOIN webpage_activity ON sidewalk_user.user_id = webpage_activity.user_id
            WHERE email LIKE 'anonymous@%'
                AND username <> 'anonymous'
                AND activity LIKE 'Visit_%'
        )
);

-- delete from sidewalk_user
DELETE
FROM sidewalk_user
WHERE email LIKE 'anonymous@%'
    AND username <> 'anonymous'
    AND sidewalk_user.user_id NOT IN (
        SELECT sidewalk_user.user_id FROM sidewalk_user
        INNER JOIN webpage_activity ON sidewalk_user.user_id = webpage_activity.user_id
        WHERE email LIKE 'anonymous@%'
            AND username <> 'anonymous'
            AND activity LIKE 'Visit_%'
    );
```

**Comment by misaugstad** [comment]
I've rewritten the queries here to be much faster. I also added a query to delete the corresponding entries in the `webpage_activity` that I had forgotten to delete. I haven't been able to speed up the final query though. Still working on it. It works for small tables just fine, but it's not finishing in a timely manner on our Seattle server.

```
-- delete user_role
DELETE
FROM user_role
WHERE user_id IN (
    SELECT sidewalk_user.user_id
    FROM (
        SELECT user_id, COUNT(*) AS count
        FROM webpage_activity
        GROUP BY user_id
    ) "activity_count"
    INNER JOIN sidewalk_user ON sidewalk_user.user_id = activity_count.user_id
    INNER JOIN webpage_activity ON activity_count.user_id = webpage_activity.user_id
    WHERE count = 1
        AND email LIKE 'anonymous@%'
        AND username <> 'anonymous'
        AND activity = 'AnonAutoSignUp'
);

-- delete user_password_info
DELETE
FROM user_password_info
    USING login_info
WHERE user_password_info.login_info_id = login_info.login_info_id
    AND provider_key IN (
        SELECT sidewalk_user.email
        FROM (
            SELECT user_id, COUNT(*) AS count
            FROM webpage_activity
            GROUP BY user_id
        ) "activity_count"
        INNER JOIN sidewalk_user ON sidewalk_user.user_id = activity_count.user_id
        INNER JOIN webpage_activity ON activity_count.user_id = webpage_activity.user_id
        WHERE count = 1
            AND email LIKE 'anonymous@%'
            AND username <> 'anonymous'
            AND activity = 'AnonAutoSignUp'
    );

-- delete user_login_info
DELETE
FROM user_login_info
WHERE user_id IN (
    SELECT sidewalk_user.user_id
    FROM (
        SELECT user_id, COUNT(*) AS count
        FROM webpage_activity
        GROUP BY user_id
    ) "activity_count"
    INNER JOIN sidewalk_user ON sidewalk_user.user_id = activity_count.user_id
    INNER JOIN webpage_activity ON activity_count.user_id = webpage_activity.user_id
    WHERE count = 1
        AND email LIKE 'anonymous@%'
        AND username <> 'anonymous'
        AND activity = 'AnonAutoSignUp'
);

-- delete login_info
DELETE
FROM login_info
WHERE provider_key IN (
    SELECT sidewalk_user.email
    FROM (
        SELECT user_id, COUNT(*) AS count
        FROM webpage_activity
        GROUP BY user_id
    ) "activity_count"
    INNER JOIN sidewalk_user ON sidewalk_user.user_id = activity_count.user_id
    INNER JOIN webpage_activity ON activity_count.user_id = webpage_activity.user_id
    WHERE count = 1
        AND email LIKE 'anonymous@%'
        AND username <> 'anonymous'
        AND activity = 'AnonAutoSignUp'
);

-- delete user_stat
DELETE
FROM user_stat
WHERE user_id IN (
    SELECT sidewalk_user.user_id
    FROM (
        SELECT user_id, COUNT(*) AS count
        FROM webpage_activity
        GROUP BY user_id
    ) "activity_count"
    INNER JOIN sidewalk_user ON sidewalk_user.user_id = activity_count.user_id
    INNER JOIN webpage_activity ON activity_count.user_id = webpage_activity.user_id
    WHERE count = 1
        AND email LIKE 'anonymous@%'
        AND username <> 'anonymous'
        AND activity = 'AnonAutoSignUp'
);

-- delete from webpage_activity
DELETE
FROM webpage_activity
WHERE user_id IN (
    SELECT sidewalk_user.user_id
    FROM (
        SELECT user_id, COUNT(*) AS count
        FROM webpage_activity
        GROUP BY user_id
    ) "activity_count"
    INNER JOIN sidewalk_user ON sidewalk_user.user_id = activity_count.user_id
    INNER JOIN webpage_activity ON activity_count.user_id = webpage_activity.user_id
    WHERE count = 1
        AND email LIKE 'anonymous@%'
        AND username <> 'anonymous'
        AND activity = 'AnonAutoSignUp'
);

-- delete from sidewalk_user
DELETE
FROM sidewalk_user
USING sidewalk_user AS user_temp
LEFT JOIN webpage_activity ON user_temp.user_id = webpage_activity.user_id
WHERE sidewalk_user.user_id = user_temp.user_id
    AND webpage_activity_id IS NULL
    AND sidewalk_user.username <> 'anonymous';
```

**Comment by misaugstad** [comment]
So I think I know what the cause of this is and what the solution should approximately be. We somehow end up in this loop where there are credentials stored in the session cookie that don't match any user in our database. Since the person fails to authenticate, they are sent to the endpoint that creates an anon account for them. The account credentials are created in the database, but they aren't replacing the credentials in the session cookie. But our code thinks everything went fine, so it redirects back to where the user was trying to go, and the same thing happens because the old credentials are still stored. So you know we do this 10 or 20 times or something before the browser decides we've done too many redirects and stops us. Which is 10 to 20 bogus anon accounts.

I think that in the anon sign up code we just need to check if there are already credentials in the session cookie and if so, discard them. This might require a second redirect: one to discard the old credentials, and one to add the new ones. Either way, I think this should fix the issue going forward.

**Comment by jonfroehlich** [comment]
Did you end up fixing this @misaugstad? I seem to remember maybe @ThatOneGoat also worked on this?

**Comment by misaugstad** [comment]
I had hoped that we had fixed the issue, but I am still seeing these accounts popping up, though much less frequently than before. I'm still going in once per month and clearing the extra accounts.

**Comment by jonfroehlich** [comment]
OK, let's keep this open then.

**Comment by misaugstad** [comment]
I have not seen new anon accounts like this for the past 6 months, so this seems to have been resolved! I'm not 100% sure why it's not longer an issue, but I wasn't sure how it was even still happening to begin with :upside_down_face:
